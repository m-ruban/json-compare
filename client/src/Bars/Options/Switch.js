import React from 'react'

const Swith = ({isCompare, toggleCompare}) => {
  return (
    <>
      <span className={`parse${isCompare ? ' pointer' : ''}`} onClick={isCompare ? toggleCompare : null}>parse</span>
      <span className="divider">&</span>
      <span className={`compare${!isCompare ? ' pointer' : ''}`} onClick={!isCompare ? toggleCompare : null}>compare</span>
    </>
  )
}

export default Swith

