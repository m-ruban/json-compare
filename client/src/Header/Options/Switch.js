import React from 'react';
import PropTypes from 'prop-types';

const Switch = ({ isCompare, toggleCompare }) => {
  return (
    <>
      <span
        className={`parse${isCompare ? ' pointer' : ''}`}
        onClick={isCompare ? toggleCompare : null}
      >
        parse
      </span>
      <span className="divider">&</span>
      <span
        className={`compare${!isCompare ? ' pointer' : ''}`}
        onClick={!isCompare ? toggleCompare : null}
      >
        compare
      </span>
    </>
  );
};

Switch.propTypes = {
  isCompare: PropTypes.bool,
  toggleCompare: PropTypes.func
};

export default Switch;
