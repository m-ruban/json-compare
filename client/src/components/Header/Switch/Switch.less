.switch {
  font-weight: 500;

  &_parse {
    color: #509e62;
  }

  &_compare {
    color: #3c7775;
  }

  &__divider {
    font-weight: 500;
    margin: 0 0.3rem;
  }

  &_pointer:hover {
    cursor: pointer;
    text-decoration: underline;
  }

  &:not(&_pointer) {
    text-decoration: underline;
  }
}
