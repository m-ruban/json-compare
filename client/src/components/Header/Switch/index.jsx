import React, { useCallback, Fragment } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import withConsumer from 'hoc/withConsumer';
import trl from 'modules/translation';

const Switch = ({ isCompare, toggleCompare }) => {
    const onParseClick = useCallback(() => {
        if (!isCompare) {
            return;
        }
        toggleCompare();
    }, [isCompare, toggleCompare]);
    const onCompareClick = useCallback(() => {
        if (isCompare) {
            return;
        }
        toggleCompare();
    }, [isCompare, toggleCompare]);
    return (
        <Fragment>
            <span className={classNames('parse', { pointer: isCompare })} onClick={onParseClick}>
                {trl('Header.Switch.parse')}
            </span>
            <span className="divider">&</span>
            <span className={classNames('compare', { pointer: !isCompare })} onClick={onCompareClick}>
                {trl('Header.Switch.compare')}
            </span>
        </Fragment>
    );
};

Switch.propTypes = {
    isCompare: PropTypes.bool,
    toggleCompare: PropTypes.func,
};

export default withConsumer(Switch);
