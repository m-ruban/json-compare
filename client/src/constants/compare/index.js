export const COMPARE_DIFF = 'diff';
export const COMPARE_REQ = 'req';
