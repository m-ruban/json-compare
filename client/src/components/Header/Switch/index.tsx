import classNames from 'classnames';
import React from 'react';

import withConsumer from 'hoc/withConsumer';
import trl from 'modules/translation';

import 'components/Header/Switch/Switch.less';

const Switch = ({ isCompare, toggleCompare }: { isCompare: boolean; toggleCompare: () => void; }) => {
    const onParseClick = React.useCallback(() => {
        if (!isCompare) {
            return;
        }
        toggleCompare();
    }, [isCompare, toggleCompare]);

    const onCompareClick = React.useCallback(() => {
        if (isCompare) {
            return;
        }
        toggleCompare();
    }, [isCompare, toggleCompare]);

    return (
        <>
            <span
                className={classNames('switch', 'switch_parse', { switch_pointer: isCompare })}
                onClick={onParseClick}
            >
                {trl('Header.Switch.parse')}
            </span>
            <span className="switch__divider">{trl('Header.Switch.and')}</span>
            <span
                className={classNames('switch', 'switch_compare', { switch_pointer: !isCompare })}
                onClick={onCompareClick}
            >
                {trl('Header.Switch.compare')}
            </span>
        </>
    );
};

export default withConsumer(Switch);
