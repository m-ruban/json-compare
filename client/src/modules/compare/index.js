import { COMPARE_EQ, COMPARE_DIFF, COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT, EQUALITY_TYPE_NOT_STRICT, EQUALITY_TYPE_IGNORE } from 'constants/equality';

let equalityTypes, state;
export function equalAlerts(left, right, type) {
    state = new Map();
    equalityTypes = type;
    compare(left, right);
    return state;
}

const findByPath = (obj, paths) => {
    let current = obj;
    for (let i = 0; i < paths.length; ++i) {
        const path = paths[i];
        if (current[path] === undefined) {
            return undefined;
        } else {
            current = current[path];
        }
    }
    return current;
};

const compareValue = (lvalue, rvalue, type) => {
    const op = {
        [EQUALITY_TYPE_NOT_STRICT]: (a, b) => a == b,
        [EQUALITY_TYPE_STRICT]: (a, b) => a === b,
    };
    let res;
    if (!rvalue) {
        res = COMPARE_REQ;
    } else {
        if (typeof lvalue !== 'object') {
            res = op[type](lvalue, rvalue) ? COMPARE_EQ : COMPARE_DIFF;
        }
    }
    return res;
};

const compare = (left, right, keys = []) => {
    for (const prop in left) {
        if (!left.hasOwnProperty(prop)) continue;
        const value = left[prop],
            type = typeof value,
            path = keys.slice();

        path.push(prop);
        const rvalue = findByPath(right, path);

        let res;
        switch (equalityTypes) {
            case EQUALITY_TYPE_IGNORE:
                res = rvalue ? COMPARE_EQ : COMPARE_REQ;
                break;
            default:
                res = compareValue(value, rvalue, equalityTypes);
                break;
        }

        if (res !== COMPARE_EQ) {
            state.set(path.join('-'), { res: res, scalar: type !== 'object' });
        }
        if (type === 'object') {
            compare(value, right, path);
        }
    }
};
