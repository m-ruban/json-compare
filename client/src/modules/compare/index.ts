import { COMPARE_DIFF, COMPARE_EQ, COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_IGNORE, EQUALITY_TYPE_NOT_STRICT, EQUALITY_TYPE_STRICT } from 'constants/equality';
import { IAlert } from 'constants/interfaces';
import {string} from 'prop-types';

let equalityTypes: string;
let state;
export function equalAlerts(left: any[] | object, right: any[] | object, type: string): Map<string, IAlert> {
    state = new Map();
    equalityTypes = type;
    compare(left, right);
    return state;
}

const findByPath = (obj: any, paths: string[]): any | undefined => {
    let current: any = obj;
    for (const path of paths) {
        if (current[path] === undefined) {
            return undefined;
        } else {
            current = current[path];
        }
    }
    return current;
};

const compareValue = (lvalue: any, rvalue: any, type: string): string => {
    const op: object = {
        [EQUALITY_TYPE_NOT_STRICT]: (a, b) => a == b,// tslint:disable-line
        [EQUALITY_TYPE_STRICT]: (a, b) => a === b,
    };
    let res: string;
    if (!rvalue) {
        res = COMPARE_REQ;
    } else {
        if (typeof lvalue !== 'object') {
            res = op[type](lvalue, rvalue) ? COMPARE_EQ : COMPARE_DIFF;
        }
    }
    return res;
};

const compare = (left: any, right: any, keys: string[] = []): void => {
    for (const prop in left) {
        if (!left.hasOwnProperty(prop)) {
            continue;
        }
        const value: any = left[prop];
        const type: string = typeof value;
        const path: string[] = keys.slice();
        path.push(prop);
        const rvalue: any = findByPath(right, path);

        let res: string;
        if (equalityTypes === EQUALITY_TYPE_IGNORE) {
            res = rvalue ? COMPARE_EQ : COMPARE_REQ;
        } else {
            res = compareValue(value, rvalue, equalityTypes);
        }

        if (res !== COMPARE_EQ) {
            state.set(path.join('-'), { res, scalar: type !== 'object' });
        }
        if (type === 'object') {
            compare(value, right, path);
        }
    }
};

export const getType = (value: any): string => {
    let type: string = typeof value;
    if (type === 'object') {
        if (Array.isArray(value)) {
            type = 'array';
        }
        if (value === null) {
            type = 'null';
        }
    }
    return type;
};
