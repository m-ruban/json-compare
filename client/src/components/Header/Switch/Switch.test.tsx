import React from 'react';

import { render } from 'enzyme';

import Switch from 'components/Header/Switch';
import { Provider } from 'store/Context';
import compare from 'tests/mocks/compare';

describe('<Switch />', () => {
    it('Renders correctly', () => {
        const output = render(
            <Provider value={compare}>
                <Switch />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
