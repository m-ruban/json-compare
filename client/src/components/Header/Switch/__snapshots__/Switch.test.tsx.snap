// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> Renders correctly 1`] = `
Array [
  <span
    class="switch switch_parse switch_pointer"
  >
    parse
  </span>,
  <span
    class="switch__divider"
  >
    &
  </span>,
  <span
    class="switch switch_compare"
  >
    compare
  </span>,
]
`;
